testflinger_agent/__init__.py

```diff
@@ -26,6 +31,87 @@
 logger = logging.getLogger(__name__)
 
 
+class ReqBufferTimer(Timer):
+    """Requests buffer flush"""
+
+    def run(self):
+        """Loop timer"""
+        while not self.finished.wait(self.interval):
+            self.function(*self.args, **self.kwargs)
+
+
+class ReqBufferHandler(logging.Handler):
+    """Requests logging handler"""
+
+    def __init__(self, agent, server):
+        super().__init__()
+        uri = urljoin(server, "/v1/agents/data/")
+        self.url = urljoin(uri, agent)
+        self.qdepth = 100  # messages
+        self.buffer = deque([], maxlen=self.qdepth)
+        self.reqbuf_timer = None
+        self.reqbuf_interval = 10.0  # seconds
+        self._start_rb_timer()
+        # reuse socket
+        self.session = self._requests_retry()
+
+    def _requests_retry(self, retries=3):
+        """Retry api server"""
+        session = requests.Session()
+        retry = Retry(
+            total=retries,
+            read=retries,
+            connect=retries,
+            backoff_factor=0.3,
+            status_forcelist=(500, 502, 503, 504),
+        )
+        adapter = HTTPAdapter(max_retries=retry)
+        session.mount("http://", adapter)
+        session.mount("https://", adapter)
+        return session
+
+    def _start_rb_timer(self):
+        """Periodically check and send buffer"""
+        self.reqbuf_timer = ReqBufferTimer(self.reqbuf_interval, self.flush)
+        # terminate timer on exit
+        self.reqbuf_timer.daemon = True
+        self.reqbuf_timer.start()
+
+    def emit(self, record):
+        """Write logging events to buffer"""
+        if len(self.buffer) >= self.qdepth:
+            self.buffer.popleft()
+
+        self.buffer.append(record)
+
+    def flush(self):
+        """Flush and post buffer"""
+        try:
+            for record in self.buffer:
+                self.session.post(
+                    url=self.url, json=self.format(record), timeout=3
+                )
+        except Exception as e:
+            logger.exception(e)
+
+            # preserve buffer
+            if len(self.buffer) <= self.qdepth:
+                return
```

Comment:
so if we got here and the buffer is bigger than qdepth, then it falls through and loses everything in it? I don't think that's what we want. I'm not sure what the corner case is where it could have more data than the qdepth in it since you check that in emit(), is there one I'm not thinking of? otherwise, it should return unconditonally here I think so that the buffer doesn't get cleared

Reply:
Easy enough, for some reason I thought this was desired, but we can just have it continuously add to the buffer stack within emit().